### PR TITLE
Enable running tests in Electron

### PR DESCRIPTION
**Pull request by jhermsmeier**
**Connects to:** #85

### CODE REVIEW

**Comment by jviotti** [comment]
Looks good. Are you preparing another PR to run on `electron-mocha` on Windows?

File: package.json [hunk 1]
```diff
@@ -24,6 +24,7 @@
   },
   "scripts": {
     "test": "npm run lint && mocha --recursive tests/*.spec.js -R spec && node tests/e2e.js",
+    "test-electron": "electron-mocha --recursive tests/*.spec.js -R spec",
```
Comment (jviotti) [inline package.json]: Can we add `electron-mocha` as a development dependency?

File: .travis.yml [hunk 2]
```diff
@@ -1,7 +1,15 @@
...
+  - npm install --global electron electron-mocha
+  - npm install
 node_js:
   - "6"
 notifications:
   email: false
+before_script:
+  - export DISPLAY=:99.0; sh -e /etc/init.d/xvfb start
+script:
+  - npm test
```
Comment (jviotti) [inline .travis.yml]: Maybe we should name this `npm run test-node` in for symmetric purposes? If so, we can also make `npm test` run both of them.

File: package.json (hunk 1 above)
Comment (jhermsmeier) [inline package.json]: Yup, done

File: .travis.yml (hunk 2 above)
Comment (jhermsmeier) [inline .travis.yml]: Yeah, good idea – added.

File: .travis.yml [hunk 3]
```diff
@@ -1,6 +1,13 @@
 language: node_js
 before_install:
   - npm install -g npm@latest
+install:
+  - npm install --global electron
```
Comment (jviotti) [inline .travis.yml]: Can we also make `electron` a development dependency and re-use that instead? That will give us a unified way to control in which electron version this code runs on.

Comment (jhermsmeier) [inline .travis.yml]: Now that I'm thinking about it, do we really want to test this on Electron as well in this repository? Since Electron is built on Node, it shouldn't be different anyways (if we stick to whatever Node version Electron has in Etcher). Plus, `etcher-image-write` is indirectly tested in Etcher.

This is starting to feel like considerable and not quite necessary overhead to me.
What do you think?

File: .travis.yml [hunk 4]
```diff
@@ -1,6 +1,13 @@
...
+  - npm install
+before_script:
+  - export DISPLAY=:99.0; sh -e /etc/init.d/xvfb start
```
Comment (lurch) [inline .travis.yml]: Would it make sense for this to be using the same `xvfb-run` as https://github.com/resin-io/etcher/blob/master/.travis.yml#L60 ?

File: package.json [hunk 5]
```diff
@@ -23,14 +23,17 @@
     "test": "tests"
   },
   "scripts": {
-    "test": "npm run lint && mocha --recursive tests/*.spec.js -R spec && node tests/e2e.js",
+    "test": "npm run lint && npm run test-node && npm run test-electron",
+    "test-node": "mocha --recursive tests/*.spec.js -R spec && node tests/e2e.js",
+    "test-electron": "electron-mocha --recursive tests/*.spec.js -R spec",
```
Comment (lurch) [inline package.json]: Any reason for `tests/e2e.js` to be included in `test-node` but not `test-electron`? Or is that another one of my stupid questions? ;-)

File: package.json [hunk 6]
```diff
@@ -23,14 +23,17 @@
...
     "lint": "eslint lib tests bin example.js",
     "readme": "jsdoc2md --template doc/README.hbs lib/index.js > README.md"
   },
   "author": "Juan Cruz Viotti <juan@resin.io>",
   "license": "Apache-2.0",
   "devDependencies": {
     "drivelist": "^5.0.6",
```
Comment (lurch) [inline package.json]: Not related to this PR specifically, but would it make sense for this `drivelist` version to be updated?

File: .travis.yml (hunk 3 above)
Comment (lurch) [inline .travis.yml]: I thought that Electron tended to be running on a slightly different version to "mainstream" NodeJS?
(and I guess we should probably target the same version of Electron in here as Etcher is using?)

Comment (jhermsmeier) [inline .travis.yml]: As far as I know Electron is tracking latest stable node (currently 7.x):
```
> process.versions.node
< "7.4.0"
```
The different version thing is the v8 ABI, which usually is only of concern for native bindings.

Comment (jviotti) [inline .travis.yml]: The issue is that Electron patches the version of node, and even monkey patches some of its standard libraries (most notably `fs`, which is heavily used here), therefore I think it is beneficial to run tests on Electron (we shouldn't drop that).

My proposal for including `electron` as a development dependency is to be able to set it to the current Electron version Etcher is targeting (which is 1.4.4 at the time of this writing). You can of course do something like `npm install -g electron@1.4.4` however we then have to write instructions asking people to install electron globally before being able to run the tests.

File: package.json (hunk 6 above)
Comment (jhermsmeier) [inline package.json]: Yup, updated in https://github.com/resin-io-modules/etcher-image-write/pull/88

File: package.json (hunk 5 above)
Comment (jhermsmeier) [inline package.json]: They're currently not run with mocha, but I guess we could port them? /cc @jviotti

Comment (jviotti) [inline package.json]: Damn, I just realized that we're not making those run on Electron, which are the main ones we're interested in :( Let's get this merge as it is, and port `e2e.js` to mocha.

Comment (jhermsmeier) [inline package.json]: I ported them already (just pushed), but discovered quite a few of those tests are actually broken (but passed regardless of outcome) – also fixed them, but the bmaps test case will need an assert or two more.

File: package.json [hunk 7]
```diff
@@ -37,7 +37,6 @@
     "jsdoc-to-markdown": "^2.0.1",
     "mocha": "^3.2.0",
     "mochainon": "^1.0.0",
-    "wary": "^1.1.0"
```
Comment (lurch) [inline package.json]: Does `mocha` also write to temporary files, the way that `wary` did?

File: .travis.yml [hunk 8]
```diff
@@ -2,7 +2,7 @@ language: node_js
 before_install:
   - npm install -g npm@latest
 install:
-  - npm install --global electron
+  - npm install --global electron@1.4.4
```
Comment (lurch) [inline .travis.yml]: I'm probably missing something obvious, but what's wrong with making electron a dev-dependency in `package.json` as @jviotti suggested?

Comment (jhermsmeier) [inline .travis.yml]: @jviotti @lurch To be honest I'd rather have Electron as an "external" dependency and mention it in the README, as having several repositories with Electron as a dev dependency tends to needlessly fill up disk space with the 140MB+ Electron comes with, and in most cases `npm run test-node` locally should cut it

Comment (jviotti) [inline .travis.yml]: It was a suggestion, but not a strong one, I'm happy with both ways.

Comment (jhermsmeier) [inline .travis.yml]: Aye, I added notes on testing with Electron in the README now

File: package.json (hunk 7 above)
Comment (jhermsmeier) [inline package.json]: No, but we don't need that in this case, as the output shouldn't change anyways, and if it does, we'd want to be able to inspect it (which is hard to do with wary).

File: .travis.yml (hunk 4 above)
Comment (jhermsmeier) [inline .travis.yml]: Updated to use `xvfb-run`

File: package.json (hunk 7 above)
Comment (lurch) [inline package.json]: Fair enough. I had a local-branch in progress with changes related to these image-writing tests, but I'm happy to rebase after this PR has been merged. (and the removal of `wary` will actually make that easier)

> we don't need that in this case, as the output shouldn't change anyways

AFAIK the output-changing shouldn't matter (my local-branch actually removes the output files from git and creates and deletes them on-demand); I'd assumed `wary` was used so that all the "big writes" happen in a temporary directory, rather than on the local filesystem? \*shrug\*

Comment (jhermsmeier) [inline package.json]: > AFAIK the output-changing shouldn't matter

Well, if the output != expected output, it does. And in this case it would go undetected, as the tests didn't explicitly check, and there was no way to inspect the output post-mortem.

> so that all the "big writes" happen in a temporary directory, rather than on the local filesystem?

I'm not sure I follow; Temporary files / directories are on the local file system, just not in the repository.

Comment (lurch) [inline package.json]: > as the tests didn't explicitly check

There's lots of different image-tests, and the ones I was looking at did verify that the output matched expected - I guess we must have been looking at different tests.

> Temporary files / directories are on the local file system

I was thinking maybe `/tmp` was in a RAM-disk, and the local repository was on a HDD with slow write speeds. But that's such a rare corner-case it's probably not worth worrying about :)

**Comment by jhermsmeier** [comment]
@jviotti Any updates on this one? I think I've mended your requested changes :)

**Comment by jviotti** [comment]
Thanks for the ping, looks like I forgot to approve back then?
